src/assets/main.js

```diff
@@ -0,0 +1,119 @@
+      }
+    } else if (e.key == "ArrowDown") {
+      this.setState({selectedResultIndex: Math.min(
+        this.state.selectedResultIndex + 1,
+        this.state.searchResults.length - 1
+      )});
+    } else if (e.key == "ArrowUp") {
+      this.setState({selectedResultIndex: Math.max(
+        this.state.selectedResultIndex - 1,
+        -1
+      )});
+    } else if (e.key == "Enter") {
+      if (this.state.selectedResultIndex != -1) {
+        window.location = this.state.searchResults[this.state.selectedResultIndex].id;
+      }
+    }
+  }
+
+  handleChange(query) {
+    if (!this.state.disabled) {
+      let searchResults = this.searchIndex.search(query);
+
+      //if no results, try suggested search instead
+      if (searchResults.length == 0) {
+        const suggestions = this.searchIndex.autoSuggest(query);
+        if (suggestions.length > 0) {
+          const firstSuggestionTerm = suggestions[0].terms[0];
+          searchResults = this.searchIndex.search(firstSuggestionTerm);
+        }
+      }
+
+      this.setState({
+        query,
+        searchResults,
+        selectedResultIndex: -1
+      });
+    }
+  }
+
+  componentDidMount() {
+    fetch("/assets/search-index.json")
+      .then(res => res.text())
+      .then(indexJson => {
+        this.searchIndex = MiniSearch.loadJSON(indexJson, {
+          idField: "path",
+          fields: ["title", "text"],
+          storeFields: ["title"],
+          searchOptions: {
+            boost: {title: 3, keywords: 2},
+            fuzzy: 0.2
+          }
+        });
+        this.setState({disabled: false});
+      });
+    window.addEventListener("keydown", (e) => {
+      if (e.key == "s" && this.inputRef && document.activeElement !== this.inputRef) {
+        this.inputRef.focus();
+        e.preventDefault();
+      }
+    });
+  }
+
+  render() {
+    return html`
+      <input
+        ref=${(el) => this.inputRef = el}
+        class="search-input ${!!this.state.query ? "nonempty" : ""}"
+        type="text"
+        placeholder="Search c20..."
+        disabled=${this.state.disabled}
+        value=${this.state.query}
+        onInput=${(e) => this.handleChange(e.target.value)}
+        onKeyDown=${this.handleKeyDown}
+      />
+      ${this.state.query != "" && html`
+        <nav class="search-results">
+          <button class="clear-button" onClick=${() => this.handleChange("")}>Close</button>
+          <h1>Search results</h1>
+          ${this.state.searchResults.length > 0 ? html`
+            <ul>
+              ${this.state.searchResults.map((result, i) => {
+                const isSelected = i == this.state.selectedResultIndex;
+                return html`<li class=${isSelected ? "selected" : ""}><a href=${result.id}>${result.title}</a></li>`
+              })}
+            </ul>
+          ` : html`
+            <p>No results found for <strong>${this.state.query}</strong></p>
+          `}
+        </nav>
+      `}
+    `;
```

Comment:
The level of inline here is genuinely terrifying, but also deeply interesting that you can do that with \`\` strings in js.

Maybe a few comments inside of the function inlines would do good here. Because I get what it's doing. But it took me a few minutes.

Reply:
I will say that the deep nesting of HTML to JS to HTML is pretty par for the course (for better or worse) for the JSX flavour of JS that this emulates and writing UI components. I agree some comments would help here, and there are some cases where I have HTML attributes defined inline which could be set at the start of the function.